designs/component-config.md

```diff
@@ -0,0 +1,318 @@
+- Provide an embeddable `ControllerConfiguration` type for easily authoring `ComponentConfig` types
+- Provide an `DefaultControllerConfig` to make the switching easier for clients
+
+### Non-Goals/Future Work
+
+- `kubebuilder` implementation and design in another PR
+- Changing the default `controller-runtime` implementation
+- Dynamically reloading `ComponentConfig` object
+- Providing `flags` interface and overrides
+
+## Proposal
+
+The `ctrl.Manager` _SHOULD_ support loading configurations from `ComponentConfig` like objects. 
+An interface for that object with getters for the specific configuration parameters is created to bridge existing patterns.
+
+Without breaking the current `ctrl.NewManager` which uses an exported `ctrl.Options{}` the `manager.go` can expose a new func, `NewFromComponentConfig()` this would be able to loop through the getters to populate an internal `ctrl.Options{}` and pass that into `New()`.
+
+```golang
+//pkg/manager/manager.go
+
+// ManagerConfiguration defines what the ComponentConfig object for ControllerRuntime needs to support
+type ManagerConfiguration interface {
+	GetSyncPeriod() *time.Duration
+
+	GetLeaderElection() bool
+	GetLeaderElectionNamespace() string
+	GetLeaderElectionID() string
+
+	GetLeaseDuration() *time.Duration
+	GetRenewDeadline() *time.Duration
+	GetRetryPeriod() *time.Duration
+
+	GetNamespace() string
+	GetMetricsBindAddress() string
+	GetHealthProbeBindAddress() string
+
+	GetReadinessEndpointName() string
+	GetLivenessEndpointName() string
+
+	GetPort() int
+	GetHost() string
+
+	GetCertDir() string
+}
+
+func NewFromComponentConfig(config *rest.Config, scheme *runtime.Scheme, filename string, managerconfig ManagerConfiguration) (Manager, error) {
+	codecs := serializer.NewCodecFactory(scheme)
+    if err := decodeComponentConfigFileInto(codecs, filename, managerconfig); err != nil {
+		
+	}
+	options := Options{}
+
+	if scheme != nil {
+		options.Scheme = scheme
+	}
+
+	// Loop through getters
+	if managerconfig.GetLeaderElection() {
+			managerconfig.GetLeaderElection()
+	}
+	// ...
+
+	return New(config, options)
+}
+```
+
+#### ComponentConfig Load Order
+
+![ComponentConfig Load Order](/designs/images/component-config-load.png)
+
+#### Embeddable ComponentConfig Type
+
+To make this easier for Controller authors `controller-runtime` can expose a set of `config.ControllerConfiguration` type that can be embedded similar to the way that `k8s.io/apimachinery/pkg/apis/meta/v1` works for `TypeMeta` and `ObjectMeta` these could live in `pkg/api/config/v1alpha1/types.go`. See the `DefaultComponentConfig` type below for and example implementation.
+
+```golang
+// pkg/api/config/v1alpha1/types.go
+package v1alpha1
+
+import (
+	"time"
+
+	configv1alpha1 "k8s.io/component-base/config/v1alpha1"
+)
+
+// ControllerConfiguration defines the embedded RuntimeConfiguration for controller-runtime clients.
+type ControllerConfiguration struct {
```

Comment:
Maybe I'm missing something, but should this struct include `Namespace` to align with the [`ManagerConfiguration` interface](https://github.com/kubernetes-sigs/controller-runtime/pull/818/files#diff-52330260780d994c6c68211c4961e13cR88)?

Also, controller-runtime supports [multi-namespace caches](https://github.com/kubernetes-sigs/controller-runtime/blob/a11a908d91e0c9aeabad65c8788ed4616a864ced/pkg/cache/multi_namespace_cache.go#L42). Do we expect users to be able to use ComponentConfig for this use case, or will we require users to set [`Options.NewCache`](https://github.com/kubernetes-sigs/controller-runtime/blob/a11a908d91e0c9aeabad65c8788ed4616a864ced/pkg/manager/manager.go#L180) manually.

Reply:
Good catch, yes it should. I glanced over that since `configv1alpha1.LeaderElectionConfiguration` has `.ResourceNamespace` but yes we should support `.Namespace` on this type.

WRT multi-namespace caches, I would hope we can support this but I'm not very familiar with the current way this is used if you have more info that would be helpful.

Reply:
For single namespaces or all namespaces, using `opts.Namespace` does the right thing. For multiple explicit namespaces, it is necessary to use:

```go
namespaces := []string{"ns1", "ns2"}
opts.NewCache = cache.MultiNamespacedCacheBuilder(namespaces)
```

Reply:
I feel like this should be possible as long as we have a `Namespaces` field in the type. 

What do you think about doing this as an addition after this is initial design is implemented? Since it already requires developers to add custom implementations it would be the same after this first implementation then maybe you and I can figure the edge cases we might run into and we can add that feature after?

Reply:
Yep, this could definitely be done as a non-breaking change as a new feature. Sounds good :+1: